File: packages/slot-math-kernels/src/slot_math_kernels/cluster_pays.py
"""W244 wave 21 — closed-form analytical model for `cluster_pays`.

Industry pattern (NetEnt Aloha Cluster Pays, Pragmatic Sweet Bonanza,
Pragmatic Gates of Olympus, BTG Money Cart, ELK Mystery Mish-Mash):

  Cluster-pays grid game
  ----------------------
    Grid (e.g. 5×3, 6×4, 7×7) with N symbols. After drop, BFS flood-fill
    finds CONNECTED REGIONS of same symbol (4-way orthogonal adjacency,
    or 8-way diagonal in some variants).

    A cluster of size ≥ `min_cluster_size` (typical 5) PAYS according to
    a per-symbol pay ladder indexed by cluster size:

      pay[symbol][cluster_size] = value × bet

    Multiple clusters per spin sum independently.

  Closed-form RTP contribution
  ----------------------------
    Closed-form for cluster size distribution from first principles
    (site percolation on a finite grid) is intractable in general. The
    industry-standard auditable approach is:

      1. Operator supplies an EMPIRICAL `cluster_size_distribution`
         per symbol, derived from PAR / MC ground truth.
      2. Kernel aggregates: RTP = sum_{sym, size}(
             p[sym][size] × pay[sym][size]
         )

    This shifts the "math complexity" boundary into the validated PAR
    data (where it belongs for regulator audit) and keeps the kernel
    a deterministic, auditable transformation.

  Output expectation per spin
  ---------------------------
    E[pay × bet | one spin] = sum over (symbol, cluster_size) of
        cluster_count_distribution[sym][size] × pay[sym][size]

    where `cluster_count_distribution[sym][size]` is the EXPECTED NUMBER
    of clusters of size `size` of symbol `sym` per spin (NOT probability;
    multiple clusters per spin can exist).

Pure-stdlib. Used by:
  * `tools.math_dsl.compile` for IR emission
  * `tools/build_cluster_pays_kernel.py` for acceptance artefact
  * `tools/tests/test_w244_cluster_pays_kernel.py` for closed-form pin
"""
from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class ClusterPaysParams:
    """Closed-form model inputs.

    Cluster distributions and pay tables are indexed by symbol id (str)
    then by cluster size (int).
    """
    # E[cluster_count_per_spin]: {symbol_id: {cluster_size: expected_count}}
    # Sourced from PAR or empirical MC; kernel does NOT compute these.
    cluster_count_distribution: dict[str, dict[int, float]]
    # Pay ladder: {symbol_id: {cluster_size: pay_x_bet}}
    pay_table: dict[str, dict[int, float]]
    # Minimum cluster size to qualify for pay (typical 5)
    min_cluster_size: int = 5
    # Grid topology for audit display
    grid_rows: int = 7
    grid_cols: int = 7
    # Adjacency rule for audit display ("4-way" / "8-way")
    adjacency: str = "4-way"

    def __post_init__(self):
        if self.min_cluster_size < 1:
            raise ValueError("min_cluster_size must be ≥ 1")
        if self.grid_rows < 1 or self.grid_cols < 1:
            raise ValueError("grid dimensions must be ≥ 1")
        if self.adjacency not in {"4-way", "8-way"}:
            raise ValueError(
                f"adjacency must be '4-way' or '8-way', got {self.adjacency!r}"
            )
        if not self.cluster_count_distribution:
            raise ValueError("cluster_count_distribution must be non-empty")
        if not self.pay_table:
            raise ValueError("pay_table must be non-empty")
        # Validate non-negative counts + pays
        for sym, dist in self.cluster_count_distribution.items():
            for size, cnt in dist.items():
                if size < 1:
                    raise ValueError(
                        f"cluster_count[{sym!r}] size {size} must be ≥ 1"
                    )
                if cnt < 0:
                    raise ValueError(
                        f"cluster_count[{sym!r}][{size}] must be ≥ 0"
                    )
        for sym, table in self.pay_table.items():
            for size, pay in table.items():
                if size < 1:
                    raise ValueError(
                        f"pay_table[{sym!r}] size {size} must be ≥ 1"
                    )
                if pay < 0:
                    raise ValueError(
                        f"pay_table[{sym!r}][{size}] must be ≥ 0"
                    )


def expected_pay_per_spin(params: ClusterPaysParams) -> float:
    """E[pay × bet | one spin].

    Sum over (symbol, cluster_size ≥ min_cluster_size) of
        cluster_count_distribution[sym][size] × pay_table[sym][size]

    Missing combinations contribute 0. Cluster sizes below threshold
    are zeroed out (industry-standard "no-pay below min" rule).
    """
    total = 0.0
    for sym, dist in params.cluster_count_distribution.items():
        sym_pay = params.pay_table.get(sym, {})
        for size, cnt in dist.items():
            if size < params.min_cluster_size:
                continue
            pay = sym_pay.get(size, 0.0)
            total += cnt * pay
    return total


def cluster_pays_rtp(params: ClusterPaysParams) -> dict:
    """Per-base-spin RTP + per-symbol audit breakdown.

    Per-spin RTP equals `expected_pay_per_spin` (cluster pays trigger
    every spin, no separate trigger probability layer — clusters are
    THE pay mechanic).
    """
    e_pay = expected_pay_per_spin(params)
    per_symbol = []
    for sym, dist in sorted(params.cluster_count_distribution.items()):
        sym_pay = params.pay_table.get(sym, {})
        sym_total = 0.0
        sizes_with_pay = []
        for size, cnt in sorted(dist.items()):
            if size < params.min_cluster_size:
                continue
            pay = sym_pay.get(size, 0.0)
            contrib = cnt * pay
            sym_total += contrib
            sizes_with_pay.append({
                "cluster_size": size,
                "expected_count_per_spin": cnt,
                "pay_x_bet": pay,
                "contribution_x_bet": contrib,
            })
        per_symbol.append({
            "symbol": sym,
            "total_contribution_x_bet": sym_total,
            "sizes": sizes_with_pay,
        })
    return {
        "rtp_contribution": e_pay,
        "grid": f"{params.grid_rows}×{params.grid_cols}",
        "adjacency": params.adjacency,
        "min_cluster_size": params.min_cluster_size,
        "per_symbol": per_symbol,
    }

File: rust-sim/src/kernels/cluster_pays.rs
//! W244.21 — cluster_pays Rust port.
//!
//! Closed-form: operator-supplied empirical cluster_count_distribution
//! × pay_table aggregation, with min_cluster_size threshold.

use serde::{Deserialize, Serialize};
use std::collections::BTreeMap;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ClusterPaysParams {
    /// {symbol: {cluster_size: expected_count_per_spin}}
    pub cluster_count_distribution: BTreeMap<String, BTreeMap<u32, f64>>,
    /// {symbol: {cluster_size: pay_x_bet}}
    pub pay_table: BTreeMap<String, BTreeMap<u32, f64>>,
    #[serde(default = "default_min_cluster_size")]
    pub min_cluster_size: u32,
    #[serde(default = "default_grid_rows")]
    pub grid_rows: u32,
    #[serde(default = "default_grid_cols")]
    pub grid_cols: u32,
    #[serde(default = "default_adjacency")]
    pub adjacency: String,
}

fn default_min_cluster_size() -> u32 { 5 }
fn default_grid_rows() -> u32 { 7 }
fn default_grid_cols() -> u32 { 7 }
fn default_adjacency() -> String { "4-way".to_string() }

impl ClusterPaysParams {
    pub fn validate(&self) -> Result<(), String> {
        if self.min_cluster_size < 1 {
            return Err("min_cluster_size must be ≥ 1".into());
        }
        if self.grid_rows < 1 || self.grid_cols < 1 {
            return Err("grid dimensions must be ≥ 1".into());
        }
        if !matches!(self.adjacency.as_str(), "4-way" | "8-way") {
            return Err(format!(
                "adjacency must be '4-way' or '8-way', got {:?}",
                self.adjacency
            ));
        }
        if self.cluster_count_distribution.is_empty() {
            return Err("cluster_count_distribution must be non-empty".into());
        }
        if self.pay_table.is_empty() {
            return Err("pay_table must be non-empty".into());
        }
        for (sym, dist) in &self.cluster_count_distribution {
            for (&size, &cnt) in dist {
                if size < 1 {
                    return Err(format!("cluster_count[{}] size must be ≥ 1", sym));
                }
                if cnt < 0.0 {
                    return Err(format!("cluster_count[{}][{}] must be ≥ 0", sym, size));
                }
            }
        }
        for (sym, table) in &self.pay_table {
            for (&size, &pay) in table {
                if size < 1 {
                    return Err(format!("pay_table[{}] size must be ≥ 1", sym));
                }
                if pay < 0.0 {
                    return Err(format!("pay_table[{}][{}] must be ≥ 0", sym, size));
                }
            }
        }
        Ok(())
    }
}

pub fn expected_pay_per_spin(p: &ClusterPaysParams) -> f64 {
    let mut total = 0.0_f64;
    for (sym, dist) in &p.cluster_count_distribution {
        if let Some(sym_pay) = p.pay_table.get(sym) {
            for (&size, &cnt) in dist {
                if size < p.min_cluster_size {
                    continue;
                }
                if let Some(&pay) = sym_pay.get(&size) {
                    total += cnt * pay;
                }
            }
        }
    }
    total
}

#[derive(Debug, Serialize)]
pub struct PerSize {
    pub cluster_size: u32,
    pub expected_count_per_spin: f64,
    pub pay_x_bet: f64,
    pub contribution_x_bet: f64,
}

#[derive(Debug, Serialize)]
pub struct PerSymbol {
    pub symbol: String,
    pub total_contribution_x_bet: f64,
    pub sizes: Vec<PerSize>,
}

#[derive(Debug, Serialize)]
pub struct ClusterPaysResult {
    pub rtp_contribution: f64,
    pub grid: String,
    pub adjacency: String,
    pub min_cluster_size: u32,
    pub per_symbol: Vec<PerSymbol>,
}

pub fn cluster_pays_rtp(p: &ClusterPaysParams) -> ClusterPaysResult {
    let e_pay = expected_pay_per_spin(p);
    let mut per_symbol: Vec<PerSymbol> = Vec::new();
    for (sym, dist) in &p.cluster_count_distribution {
        let sym_pay = p.pay_table.get(sym);
        let mut sym_total = 0.0_f64;
        let mut sizes: Vec<PerSize> = Vec::new();
        for (&size, &cnt) in dist {
            if size < p.min_cluster_size {
                continue;
            }
            let pay = sym_pay.and_then(|t| t.get(&size).copied()).unwrap_or(0.0);
            let contrib = cnt * pay;
            sym_total += contrib;
            sizes.push(PerSize {
                cluster_size: size,
                expected_count_per_spin: cnt,
                pay_x_bet: pay,
                contribution_x_bet: contrib,
            });
        }
        per_symbol.push(PerSymbol {
            symbol: sym.clone(),
            total_contribution_x_bet: sym_total,
            sizes,
        });
    }
    ClusterPaysResult {
        rtp_contribution: e_pay,
        grid: format!("{}×{}", p.grid_rows, p.grid_cols),
        adjacency: p.adjacency.clone(),
        min_cluster_size: p.min_cluster_size,
        per_symbol,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn mk_params() -> ClusterPaysParams {
        let mut dist = BTreeMap::new();
        let mut a_dist = BTreeMap::new();
        a_dist.insert(5, 0.1);
        dist.insert("A".to_string(), a_dist);
        let mut pay = BTreeMap::new();
        let mut a_pay = BTreeMap::new();
        a_pay.insert(5, 2.0);
        pay.insert("A".to_string(), a_pay);
        ClusterPaysParams {
            cluster_count_distribution: dist,
            pay_table: pay,
            min_cluster_size: 5,
            grid_rows: 5,
            grid_cols: 6,
            adjacency: "4-way".to_string(),
        }
    }

    #[test]
    fn simple_single_symbol() {
        let p = mk_params();
        let r = cluster_pays_rtp(&p);
        assert!((r.rtp_contribution - 0.2).abs() < 1e-12);
    }

    #[test]
    fn below_min_pays_zero() {
        let mut dist = BTreeMap::new();
        let mut a_dist = BTreeMap::new();
        a_dist.insert(3, 0.5);
        a_dist.insert(5, 0.1);
        dist.insert("A".to_string(), a_dist);
        let mut pay = BTreeMap::new();
        let mut a_pay = BTreeMap::new();
        a_pay.insert(3, 10.0);
        a_pay.insert(5, 2.0);
        pay.insert("A".to_string(), a_pay);
        let p = ClusterPaysParams {
            cluster_count_distribution: dist,
            pay_table: pay,
            min_cluster_size: 5,
            grid_rows: 5, grid_cols: 6,
            adjacency: "4-way".to_string(),
        };
        let r = cluster_pays_rtp(&p);
        // Only size 5 contributes (size 3 below min)
        assert!((r.rtp_contribution - 0.2).abs() < 1e-12);
    }
}
